Nil Pointer When Deleting a Plan

**Version**
v0.4.0

**Platform/Architecture**
linux-amd64

**Describe the bug**
When deleting a plan SUC will crash with a nil pointer, also the jobs won't get deleted.

**To Reproduce**
* Start clean Ubuntu vagrant machine
* `curl -sfL https://get.k3s.io | sh -`
* `curl -s "https://raw.githubusercontent.com/ kubernetes-sigs/kustomize/master/hack/install_kustomize.sh"  | bash`
* `mv kustomize /usr/local/bin`
* `kustomize build https://github.com/rancher/system-upgrade-controller | kubectl apply -f -`
* `kubectl label node agent1 plan.upgrade.cattle.io/bionic=true`
* Apply example plan: https://github.com/rancher/system-upgrade-controller/blob/master/examples/ubuntu/bionic.yaml
* Delete the plan

**Expected behavior**
* Operator should not crash
* Jobs associated with the plan should also get deleted

**Actual behavior**
Log output:
```
root@agent1:~# kubectl logs -f -n system-upgrade system-upgrade-controller-9bbf49c85-lm44s
W0421 11:55:00.376837       1 client_config.go:543] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
time="2020-04-21T11:55:00Z" level=info msg="Updating CRD plans.upgrade.cattle.io"
time="2020-04-21T11:55:00Z" level=info msg="Starting /v1, Kind=Node controller"
time="2020-04-21T11:55:00Z" level=info msg="Starting /v1, Kind=Secret controller"
time="2020-04-21T11:55:00Z" level=info msg="Starting batch/v1, Kind=Job controller"
time="2020-04-21T11:55:00Z" level=info msg="Starting upgrade.cattle.io/v1, Kind=Plan controller"
E0421 12:04:25.895701       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 125 [running]:
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1448560, 0x241b030)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1448560, 0x241b030)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/rancher/system-upgrade-controller/pkg/upgrade.(*Controller).handleSecrets.func1(0xc000428360, 0x15, 0x0, 0x24, 0xc1000000c0e, 0x40c44b)
        /go/src/github.com/rancher/system-upgrade-controller/pkg/upgrade/handle_core.go:47 +0x10c
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1.FromSecretHandlerToHandler.func1(0xc000428360, 0x15, 0x0, 0x0, 0xc000043cb0, 0x24, 0x16948a0, 0xc00011f9e0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1/secret.go:96 +0xdb
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Handlers).Handle(0xc0004113e0, 0xc000428360, 0x15, 0x0, 0x0, 0x15, 0x0, 0x0, 0xc000411100)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/handlers.go:27 +0x100
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).syncHandler(0xc0000d0700, 0xc000428360, 0x15, 0x10894e1, 0xc00012d0c8)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:171 +0x122
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).processSingleItem(0xc0000d0700, 0x13db1a0, 0xc0004680c0, 0x0, 0x0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:156 +0xf0
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).processNextWorkItem(0xc0000d0700, 0xc000244668)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:133 +0x51
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).runWorker(0xc0000d0700)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:122 +0x2b
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00021be60)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00021be60, 0x3b9aca00, 0x0, 0x13db101, 0xc000086ae0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc00021be60, 0x3b9aca00, 0xc000086ae0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).run
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:104 +0x178
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x128bfac]

goroutine 125 [running]:
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1448560, 0x241b030)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/rancher/system-upgrade-controller/pkg/upgrade.(*Controller).handleSecrets.func1(0xc000428360, 0x15, 0x0, 0x24, 0xc1000000c0e, 0x40c44b)
        /go/src/github.com/rancher/system-upgrade-controller/pkg/upgrade/handle_core.go:47 +0x10c
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1.FromSecretHandlerToHandler.func1(0xc000428360, 0x15, 0x0, 0x0, 0xc000043cb0, 0x24, 0x16948a0, 0xc00011f9e0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1/secret.go:96 +0xdb
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Handlers).Handle(0xc0004113e0, 0xc000428360, 0x15, 0x0, 0x0, 0x15, 0x0, 0x0, 0xc000411100)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/handlers.go:27 +0x100
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).syncHandler(0xc0000d0700, 0xc000428360, 0x15, 0x10894e1, 0xc00012d0c8)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:171 +0x122
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).processSingleItem(0xc0000d0700, 0x13db1a0, 0xc0004680c0, 0x0, 0x0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:156 +0xf0
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).processNextWorkItem(0xc0000d0700, 0xc000244668)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:133 +0x51
github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).runWorker(0xc0000d0700)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:122 +0x2b
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00021be60)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00021be60, 0x3b9aca00, 0x0, 0x13db101, 0xc000086ae0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc00021be60, 0x3b9aca00, 0xc000086ae0)
        /go/src/github.com/rancher/system-upgrade-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic.(*Controller).run
        /go/src/github.com/rancher/system-upgrade-controller/vendor/github.com/rancher/wrangler/pkg/generic/controller.go:104 +0x178
```

--- a/pkg/upgrade/handle_core.go
+++ b/pkg/upgrade/handle_core.go
@@ -38,6 +38,9 @@ func (ctl *Controller) handleSecrets(ctx context.Context) error {
 	plans := ctl.upgradeFactory.Upgrade().V1().Plan()
 
 	ctl.coreFactory.Core().V1().Secret().OnChange(ctx, ctl.Name, func(key string, obj *corev1.Secret) (*corev1.Secret, error) {
+		if obj == nil {
+			return obj, nil
+		}
 		planList, err := plans.Cache().List(ctl.Namespace, labels.Everything())
 		if err != nil {
 			return obj, err

--- a/pkg/upgrade/handle_upgrade.go
+++ b/pkg/upgrade/handle_upgrade.go
@@ -23,6 +23,9 @@ func (ctl *Controller) handlePlans(ctx context.Context) error {
 	// process plan events, mutating status accordingly
 	upgradectlv1.RegisterPlanStatusHandler(ctx, plans, "", ctl.Name,
 		func(obj *upgradeapiv1.Plan, status upgradeapiv1.PlanStatus) (upgradeapiv1.PlanStatus, error) {
+			if obj == nil {
+				return status, nil
+			}
 			logrus.Debugf("PLAN STATUS HANDLER: plan=%s/%s@%s, status=%+v", obj.Namespace, obj.Name, obj.ResourceVersion, status)
 			resolved := upgradeapiv1.PlanLatestResolved
 			resolved.CreateUnknownIfNotExists(obj)
@@ -58,6 +61,9 @@ func (ctl *Controller) handlePlans(ctx context.Context) error {
 	// process plan events by creating jobs to apply the plan
 	upgradectlv1.RegisterPlanGeneratingHandler(ctx, plans, ctl.apply.WithCacheTypes(jobs, nodes, secrets).WithNoDelete(), "", ctl.Name,
 		func(obj *upgradeapiv1.Plan, status upgradeapiv1.PlanStatus) (objects []runtime.Object, _ upgradeapiv1.PlanStatus, _ error) {
+			if obj == nil {
+				return objects, status, nil
+			}
 			logrus.Debugf("PLAN GENERATING HANDLER: plan=%s/%s@%s, status=%+v", obj.Namespace, obj.Name, obj.ResourceVersion, status)
 			concurrentNodeNames, err := upgradeplan.SelectConcurrentNodeNames(obj, nodes.Cache())
 			if err != nil {

